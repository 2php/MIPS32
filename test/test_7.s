# exceptions
.set noreorder
start:
  .text 0
  jal init
  nop
  div $s0, $s1
  break


init:
  
  # set exception return base (EBase)
  la $t1, 0x80000000 
  mfc0 $t0, $15
  or   $t0, $t1
  mtc0 $t0, $15
  
  # enable exception
  mfc0 $t0, $12
  or   $t0,  0b00000000000000000000000000000001
  and  $t0,~(0b00000000000000000000000000000010)
  mtc0 $t0, $12
  
  jr $ra
  nop
  
# ex_handler @ 0x80000180
.set noreorder
start:
  .text 0
  
  la $s0, 0xbfd003f8
  la $s1, 0xbfd003fc
  
  mfc0 $a0, $13
  and  $a0, 0xff
  srl  $a0, 2
  jal  write_com
  nop
  
  break
  
write_com:
  li $t1, 1
  loop:
  lw $t0, 0($s1)
  andi $t0, $t0, 1
  bne $t0, $t1, loop
  sw $a0, 0($s0)
  jr $ra
  nop



