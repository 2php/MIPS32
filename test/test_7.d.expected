booting

fetch instr @ 0x80000000 (10000000000000000000000000000000)
MMU[0x80000000] :  0x00000000
Mem[0x00000000] :  0x0c00000d (00001100000000000000000000001101)
000011 | 00000000000000000000001101
JAL      instr_index:0x000000d
R[31] <= 0x80000008 (10000000000000000000000000001000)

fetch instr @ 0x80000034 (10000000000000000000000000110100)
MMU[0x80000034] :  0x00000034
Mem[0x00000034] :  0x40086000 (01000000000010000110000000000000)
010000 | 00000 | 01000 | 01100 | 00000000 | 000
COP0     mfc0    rt: 8   SR      
R[SR      ] :  0x10000002 (00010000000000000000000000000010)
R[ 8] <= 0x10000002 (00010000000000000000000000000010)

fetch instr @ 0x80000038 (10000000000000000000000000111000)
MMU[0x80000038] :  0x00000038
Mem[0x00000038] :  0x3508ff01 (00110101000010001111111100000001)
001101 | 01000 | 01000 | 1111111100000001
ORI      rs: 8   rt: 8   imm:0xff01
R[ 8] :  0x10000002 (00010000000000000000000000000010)
alu_a <= 0x10000002 (00010000000000000000000000000010)
alu_b <= 0x0000ff01 (00000000000000001111111100000001)
alu_r :  0x1000ff03 (00010000000000001111111100000011)
R[ 8] <= 0x1000ff03 (00010000000000001111111100000011)

fetch instr @ 0x8000003c (10000000000000000000000000111100)
MMU[0x8000003c] :  0x0000003c
Mem[0x0000003c] :  0x2401fffd (00100100000000011111111111111101)
001001 | 00000 | 00001 | 1111111111111101
ADDIU    rs: 0   rt: 1   imm:0xfffd
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0xfffffffd (11111111111111111111111111111101)
alu_r :  0xfffffffd (11111111111111111111111111111101)
R[ 1] <= 0xfffffffd (11111111111111111111111111111101)

fetch instr @ 0x80000040 (10000000000000000000000001000000)
MMU[0x80000040] :  0x00000040
Mem[0x00000040] :  0x01014024 (00000001000000010100000000100100)
000000 | 01000 | 00001 | 01000 | 00000 | 100100
SPECIAL  rs: 8   rt: 1   rd: 8   sa: 0   AND    
R[ 8] :  0x1000ff03 (00010000000000001111111100000011)
R[ 1] :  0xfffffffd (11111111111111111111111111111101)
alu_a <= 0x1000ff03 (00010000000000001111111100000011)
alu_b <= 0xfffffffd (11111111111111111111111111111101)
alu_r :  0x1000ff01 (00010000000000001111111100000001)
R[ 8] <= 0x1000ff01 (00010000000000001111111100000001)

fetch instr @ 0x80000044 (10000000000000000000000001000100)
MMU[0x80000044] :  0x00000044
Mem[0x00000044] :  0x40886000 (01000000100010000110000000000000)
010000 | 00100 | 01000 | 01100 | 00000000 | 000
COP0     mtc0    rt: 8   SR      
R[ 8] :  0x1000ff01 (00010000000000001111111100000001)
R[SR      ] <= 0x1000ff01 (00010000000000001111111100000001)

fetch instr @ 0x80000048 (10000000000000000000000001001000)
MMU[0x80000048] :  0x00000048
Mem[0x00000048] :  0x3c098001 (00111100000010011000000000000001)
001111 | 00000 | 01001 | 1000000000000001
LUI      rs: 0   rt: 9   imm:0x8001
R[ 9] <= 0x80010000 (10000000000000010000000000000000)

fetch instr @ 0x8000004c (10000000000000000000000001001100)
MMU[0x8000004c] :  0x0000004c
Mem[0x0000004c] :  0x40087800 (01000000000010000111100000000000)
010000 | 00000 | 01000 | 01111 | 00000000 | 000
COP0     mfc0    rt: 8   EBase   
R[EBase   ] :  0x80000000 (10000000000000000000000000000000)
R[ 8] <= 0x80000000 (10000000000000000000000000000000)

fetch instr @ 0x80000050 (10000000000000000000000001010000)
MMU[0x80000050] :  0x00000050
Mem[0x00000050] :  0x01094025 (00000001000010010100000000100101)
000000 | 01000 | 01001 | 01000 | 00000 | 100101
SPECIAL  rs: 8   rt: 9   rd: 8   sa: 0   OR     
R[ 8] :  0x80000000 (10000000000000000000000000000000)
R[ 9] :  0x80010000 (10000000000000010000000000000000)
alu_a <= 0x80000000 (10000000000000000000000000000000)
alu_b <= 0x80010000 (10000000000000010000000000000000)
alu_r :  0x80010000 (10000000000000010000000000000000)
R[ 8] <= 0x80010000 (10000000000000010000000000000000)

fetch instr @ 0x80000054 (10000000000000000000000001010100)
MMU[0x80000054] :  0x00000054
Mem[0x00000054] :  0x40887800 (01000000100010000111100000000000)
010000 | 00100 | 01000 | 01111 | 00000000 | 000
COP0     mtc0    rt: 8   EBase   
R[ 8] :  0x80010000 (10000000000000010000000000000000)
R[EBase   ] <= 0x80010000 (10000000000000010000000000000000)

fetch instr @ 0x80000058 (10000000000000000000000001011000)
MMU[0x80000058] :  0x00000058
Mem[0x00000058] :  0x03e00008 (00000011111000000000000000001000)
000000 | 11111 | 00000 | 00000 | 00000 | 001000
SPECIAL  rs:31   rt: 0   rd: 0   sa: 0   JR     
R[31] :  0x80000008 (10000000000000000000000000001000)

fetch instr @ 0x80000008 (10000000000000000000000000001000)
MMU[0x80000008] :  0x00000008
Mem[0x00000008] :  0x16200002 (00010110001000000000000000000010)
000101 | 10001 | 00000 | 0000000000000010
BNE      rs:17   rt: 0   imm:0x0002
R[17] :  0x00000000 (00000000000000000000000000000000)
R[ 0] :  0x00000000 (00000000000000000000000000000000)

fetch instr @ 0x8000000c (10000000000000000000000000001100)
MMU[0x8000000c] :  0x0000000c
Mem[0x0000000c] :  0x0211001a (00000010000100010000000000011010)
000000 | 10000 | 10001 | 00000 | 00000 | 011010
SPECIAL  rs:16   rt:17   rd: 0   sa: 0   DIV    
Invalid Instruction
Exception (ExcCode:10)

fetch instr @ 0x80010180 (10000000000000010000000110000000)
MMU[0x80010180] :  0x00010180
Mem[0x00010180] :  0x3c10bfd0 (00111100000100001011111111010000)
001111 | 00000 | 10000 | 1011111111010000
LUI      rs: 0   rt:16   imm:0xbfd0
R[16] <= 0xbfd00000 (10111111110100000000000000000000)

fetch instr @ 0x80010184 (10000000000000010000000110000100)
MMU[0x80010184] :  0x00010184
Mem[0x00010184] :  0x361003f8 (00110110000100000000001111111000)
001101 | 10000 | 10000 | 0000001111111000
ORI      rs:16   rt:16   imm:0x03f8
R[16] :  0xbfd00000 (10111111110100000000000000000000)
alu_a <= 0xbfd00000 (10111111110100000000000000000000)
alu_b <= 0x000003f8 (00000000000000000000001111111000)
alu_r :  0xbfd003f8 (10111111110100000000001111111000)
R[16] <= 0xbfd003f8 (10111111110100000000001111111000)

fetch instr @ 0x80010188 (10000000000000010000000110001000)
MMU[0x80010188] :  0x00010188
Mem[0x00010188] :  0x3c11bfd0 (00111100000100011011111111010000)
001111 | 00000 | 10001 | 1011111111010000
LUI      rs: 0   rt:17   imm:0xbfd0
R[17] <= 0xbfd00000 (10111111110100000000000000000000)

fetch instr @ 0x8001018c (10000000000000010000000110001100)
MMU[0x8001018c] :  0x0001018c
Mem[0x0001018c] :  0x363103fc (00110110001100010000001111111100)
001101 | 10001 | 10001 | 0000001111111100
ORI      rs:17   rt:17   imm:0x03fc
R[17] :  0xbfd00000 (10111111110100000000000000000000)
alu_a <= 0xbfd00000 (10111111110100000000000000000000)
alu_b <= 0x000003fc (00000000000000000000001111111100)
alu_r :  0xbfd003fc (10111111110100000000001111111100)
R[17] <= 0xbfd003fc (10111111110100000000001111111100)

fetch instr @ 0x80010190 (10000000000000010000000110010000)
MMU[0x80010190] :  0x00010190
Mem[0x00010190] :  0x40046800 (01000000000001000110100000000000)
010000 | 00000 | 00100 | 01101 | 00000000 | 000
COP0     mfc0    rt: 4   Cause   
R[Cause   ] :  0x00000028 (00000000000000000000000000101000)
R[ 4] <= 0x00000028 (00000000000000000000000000101000)

fetch instr @ 0x80010194 (10000000000000010000000110010100)
MMU[0x80010194] :  0x00010194
Mem[0x00010194] :  0x308400ff (00110000100001000000000011111111)
001100 | 00100 | 00100 | 0000000011111111
ANDI     rs: 4   rt: 4   imm:0x00ff
R[ 4] :  0x00000028 (00000000000000000000000000101000)
alu_a <= 0x00000028 (00000000000000000000000000101000)
alu_b <= 0x000000ff (00000000000000000000000011111111)
alu_r :  0x00000028 (00000000000000000000000000101000)
R[ 4] <= 0x00000028 (00000000000000000000000000101000)

fetch instr @ 0x80010198 (10000000000000010000000110011000)
MMU[0x80010198] :  0x00010198
Mem[0x00010198] :  0x00042082 (00000000000001000010000010000010)
000000 | 00000 | 00100 | 00100 | 00010 | 000010
SPECIAL  rs: 0   rt: 4   rd: 4   sa: 2   SRL    
R[ 4] :  0x00000028 (00000000000000000000000000101000)
alu_a <= 0x00000002 (00000000000000000000000000000010)
alu_b <= 0x00000028 (00000000000000000000000000101000)
alu_r :  0x0000000a (00000000000000000000000000001010)
R[ 4] <= 0x0000000a (00000000000000000000000000001010)

fetch instr @ 0x8001019c (10000000000000010000000110011100)
MMU[0x8001019c] :  0x0001019c
Mem[0x0001019c] :  0x0c00406a (00001100000000000100000001101010)
000011 | 00000000000100000001101010
JAL      instr_index:0x000406a
R[31] <= 0x800101a4 (10000000000000010000000110100100)

fetch instr @ 0x800101a8 (10000000000000010000000110101000)
MMU[0x800101a8] :  0x000101a8
Mem[0x000101a8] :  0x24090001 (00100100000010010000000000000001)
001001 | 00000 | 01001 | 0000000000000001
ADDIU    rs: 0   rt: 9   imm:0x0001
R[ 0] :  0x00000000 (00000000000000000000000000000000)
alu_a <= 0x00000000 (00000000000000000000000000000000)
alu_b <= 0x00000001 (00000000000000000000000000000001)
alu_r :  0x00000001 (00000000000000000000000000000001)
R[ 9] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x800101ac (10000000000000010000000110101100)
MMU[0x800101ac] :  0x000101ac
Mem[0x000101ac] :  0x8e280000 (10001110001010000000000000000000)
100011 | 10001 | 01000 | 0000000000000000
LW       rs:17   rt: 8   imm:0x0000
R[17] :  0xbfd003fc (10111111110100000000001111111100)
alu_a <= 0xbfd003fc (10111111110100000000001111111100)
alu_b <= 0x00000000 (00000000000000000000000000000000)
alu_r :  0xbfd003fc (10111111110100000000001111111100)
MMU[0xbfd003fc] :  0x1fd003fc
Mem[0x1fd003fc] :  0x00000001 (00000000000000000000000000000001)
R[ 8] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x800101b0 (10000000000000010000000110110000)
MMU[0x800101b0] :  0x000101b0
Mem[0x000101b0] :  0x31080001 (00110001000010000000000000000001)
001100 | 01000 | 01000 | 0000000000000001
ANDI     rs: 8   rt: 8   imm:0x0001
R[ 8] :  0x00000001 (00000000000000000000000000000001)
alu_a <= 0x00000001 (00000000000000000000000000000001)
alu_b <= 0x00000001 (00000000000000000000000000000001)
alu_r :  0x00000001 (00000000000000000000000000000001)
R[ 8] <= 0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x800101b4 (10000000000000010000000110110100)
MMU[0x800101b4] :  0x000101b4
Mem[0x000101b4] :  0x1509fffd (00010101000010011111111111111101)
000101 | 01000 | 01001 | 1111111111111101
BNE      rs: 8   rt: 9   imm:0xfffd
R[ 8] :  0x00000001 (00000000000000000000000000000001)
R[ 9] :  0x00000001 (00000000000000000000000000000001)

fetch instr @ 0x800101b8 (10000000000000010000000110111000)
MMU[0x800101b8] :  0x000101b8
Mem[0x000101b8] :  0xae040000 (10101110000001000000000000000000)
101011 | 10000 | 00100 | 0000000000000000
SW       rs:16   rt: 4   imm:0x0000
R[16] :  0xbfd003f8 (10111111110100000000001111111000)
R[ 4] :  0x0000000a (00000000000000000000000000001010)
alu_a <= 0xbfd003f8 (10111111110100000000001111111000)
alu_b <= 0x00000000 (00000000000000000000000000000000)
alu_r :  0xbfd003f8 (10111111110100000000001111111000)
MMU[0xbfd003f8] :  0x1fd003f8
Mem[0x1fd003f8] <= 0x0a (00001010)

fetch instr @ 0x800101bc (10000000000000010000000110111100)
MMU[0x800101bc] :  0x000101bc
Mem[0x000101bc] :  0x03e00008 (00000011111000000000000000001000)
000000 | 11111 | 00000 | 00000 | 00000 | 001000
SPECIAL  rs:31   rt: 0   rd: 0   sa: 0   JR     
R[31] :  0x800101a4 (10000000000000010000000110100100)

fetch instr @ 0x800101a4 (10000000000000010000000110100100)
MMU[0x800101a4] :  0x000101a4
Mem[0x000101a4] :  0x0000000d (00000000000000000000000000001101)
000000 | 00000 | 00000 | 00000 | 00000 | 001101
SPECIAL  rs: 0   rt: 0   rd: 0   sa: 0   BREAK  
halt
